File: java/timesheet/src/main/java/timesheet/TimesheetReminderService.java
package timesheet;

import java.time.LocalDateTime;
import java.util.List;

public class TimesheetReminderService {
    private final EmployeeRepository employeeRepository;

    public TimesheetReminderService(EmployeeRepository employeeRepository) {
        this.employeeRepository = employeeRepository;
    }

    public void sendTimesheetReminder() {
        List<Worker> activeWorkers = employeeRepository.findActiveWorkers();
        activeWorkers
                .stream()
                .filter(emp -> !new TimesheetService().hasTimesheets(emp, LocalDateTime.now()))
                .map(employee -> emailFor(employee))
                .forEach(email -> new EmailSender().send(email));
    }

    private Email emailFor(Worker worker) {
        String message = String.format("You have some timesheet missing. Please hurry to send them. %s!", worker.getFirstName());
        return new Email(worker.getEmail(), "Your timesheet is missing!!", message);
    }

}

File: scala/timesheet/src/main/scala/timesheet/TimesheetReminderService.scala
package timesheet

import java.time.LocalDateTime

class TimesheetReminderService(employeeRepository: EmployeeRepository) {

  def sendTimesheetReminder(): Unit = {
    val activeWorkers = employeeRepository.findActiveWorkers()
    activeWorkers
      .filter(emp => !new TimesheetService().hasTimesheets(emp, LocalDateTime.now()))
      .map(employee => emailFor(employee))
      .foreach(email => new EmailSender().send(email))
  }

  private def emailFor(worker: Worker): Email = {
    val message = s"You have some timesheet missing. Please hurry to send them. ${worker.getFirstName}!"
    new Email(worker.getEmail, "Your timesheet is missing!!", message)
  }
}
